src/bam/collapse_reads_to_fragments/pipeline.rs

```diff
@@ -375,67 +441,57 @@ fn group_reads_by_cigar(
+                };
+                if !cigar_groups.contains_key(&cigar_tuple) {
+                    cigar_groups.insert(
+                        cigar_tuple.clone(),
+                        CigarGroup::PairedRecords {
+                            r1_recs: Vec::new(),
+                            r2_recs: Vec::new(),
+                            r1_seqids: Vec::new(),
+                            r2_seqids: Vec::new(),
+                        },
+                    );
+                }
+                match cigar_groups.get_mut(&cigar_tuple) {
+                    Some(CigarGroup::PairedRecords {
+                        r1_recs,
+                        r2_recs,
+                        r1_seqids,
+                        r2_seqids,
+                    }) => {
+                        r1_recs.push(r1_rec_entry);
+                        r2_recs.push(r2_rec_entry);
+                        r1_seqids.push(r1_rec_id);
+                        r2_seqids.push(r2_rec_id);
                     }
+                    _ => unreachable!(),
                 }
             }
             RecordStorage::SingleRecord { rec } => {
                 let rec_id = rec.rec_id;
                 let rec_entry = rec.into_rec();
-                if cigar_has_softclips(&rec_entry) {
-                    bam_skipped_writer.write(&rec_entry)?;
-                } else {
-                    let cigar_single = Cigar::Single {
-                        cigar: rec_entry.raw_cigar().to_vec(),
-                    };
-                    if !cigar_groups.contains_key(&cigar_single) {
-                        cigar_groups.insert(
-                            cigar_single.clone(),
-                            CigarGroup::SingleRecords {
-                                recs: Vec::new(),
-                                seqids: Vec::new(),
-                            },
-                        );
-                    }
-                    match cigar_groups.get_mut(&cigar_single) {
-                        Some(CigarGroup::SingleRecords { recs, seqids }) => {
-                            recs.push(rec_entry);
-                            seqids.push(rec_id);
-                        }
-                        _ => unreachable!(),
+                let cigar_single = Cigar::Single {
+                    cigar: rec_entry.raw_cigar().to_vec(),
+                };
+                if !cigar_groups.contains_key(&cigar_single) {
+                    cigar_groups.insert(
+                        cigar_single.clone(),
+                        CigarGroup::SingleRecords {
+                            recs: Vec::new(),
+                            seqids: Vec::new(),
+                        },
+                    );
+                }
+                match cigar_groups.get_mut(&cigar_single) {
+                    Some(CigarGroup::SingleRecords { recs, seqids }) => {
+                        recs.push(rec_entry);
+                        seqids.push(rec_id);
```

Comment:
code duplication?

Reply:
This is now implemented as a function of RecordStorage reducing duplication by a bit.